### ISSUE TITLE
Authorization Permission Mismatch in RPT and Permission Ticket 

### ISSUE THREAD

**Issue by sirkrypt0** [keycloak/keycloak#12512]
### Describe the bug

I was playing around with the resources feature in Keycloak and noticed a strange behavior when evaluating the authorization permissions in the access tokens of resources owned by a user other than the client itself.
While requesting an access token with grant type `urn:ietf:params:oauth:grant-type:uma-ticket` (using an access token obtained with direct grant), the permission to a resource owned by another user seem to not be granted.
However, when requesting a similar access token, but this time using a previously acquired permission ticket, the permissions seem to be granted.

Using the `evaluate` feature of the admin console, I noticed that the permissions obtained using the permission ticket match the ones evaluated in the admin console, while the other access token does not contain the expected permissions.

### Version

18.0.0

### Expected behavior

I'd expect the first access token to contain the same permissions as the second access token obtained using a permission ticket.

### Actual behavior

The permissions differ.

### How to Reproduce?

Make sure to have at least 2 users in your realm.

1. Create a new client.
    - Enable Direct Access Grants
    - Enable Service Accounts
    - Enable Authorization
2. Create a new resource owned by user1 (this can only be done through a request, at least I didn't find a way to specify a different owner than the client itself in the admin console).
   - Make sure it has some scopes assigned that we can request later on
   - Also it must have the `urn:CLIENT_ID:resources:default` type
4. Obtain the first access token
   1. Obtain an access token for user2 using direct grant
   2. Exchange the access token for a `uma-ticket` using the `urn:ietf:params:oauth:grant-type:uma-ticket` grant type.
   3. Decode the resulting token and note the authorization permissions. They don't contain the permissions for the newly created resource.
5. Obtain the second access token
  1. Obtain an access token for user2 using direct grant
  2. Exchange the direct grant for a permission ticket requesting the assigned resource scopes
  3. Decode the resulting token and note the authorization permissions again. This time, they contain the permissions for the newly created resource.

You may also adjust and use the following script to reproduce the steps above.
Note that it requires `jq` and only works as long as the resource has not been created already (as we don't get the ID otherwise).

```bash
#!/bin/bash

# base address
KC_BASE="http://localhost:8080/auth"
# realm
KC_REALM="realm"
# id of user1, the owner of the resource
KC_OWNER_ID="3019fc73-1247-4df3-ad07-298e6d66a517"
# username of user2, the one requesting access
KC_USERNAME="username"
# password of user2
KC_PASSWORD="password"
# client id of resource client
KC_CLIENT="client"
# client secret of resource client
KC_SECRET="secret"

resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/protocol/openid-connect/token" \
 -H "Content-Type: application/x-www-form-urlencoded" \
 -d "grant_type=password" \
 -d "client_id=$KC_CLIENT" -d "client_secret=$KC_SECRET" \
 -d "username=$KC_USERNAME" -d "password=$KC_PASSWORD")
userToken=$(echo $resp | jq -r .access_token)

resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/protocol/openid-connect/token" \
 -H "Content-Type: application/x-www-form-urlencoded" \
 -d "grant_type=client_credentials" \
 -d "client_id=$KC_CLIENT" -d "client_secret=$KC_SECRET")
clientToken=$(echo $resp | jq -r .access_token)

# Create resource
resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/authz/protection/resource_set" \
 -H "Content-Type: application/json" \
 -H "Authorization: Bearer $clientToken" \
 --data "{\"name\": \"demo\",\"owner\": {\"id\": \"$KC_OWNER_ID\"},\"scopes\": [{\"name\": \"read\"}],\"type\": \"urn:$KC_CLIENT:resources:default\"}"
)
echo "Created resource: $resp"

resource_id=$(echo $resp | jq -r ._id)

# Get permissions using RPT
resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/protocol/openid-connect/token" \
 -H "Content-Type: application/x-www-form-urlencoded" \
 -H "Authorization: Bearer $userToken" \
 -d "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" \
 -d "response_include_resource_name=true" \
 -d "audience=$KC_CLIENT" )

rpt=$(echo $resp | jq -r .access_token)

echo "Permissions from RPT:"
echo $rpt | jq -R 'split(".") | .[0],.[1] | @base64d | fromjson | .authorization.permissions'

# Get permissions using permission ticket
resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/authz/protection/permission" \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $userToken" \
  --data "[{\"resource_id\": \"$resource_id\",\"resource_scopes\": [\"read\"]}]" )
ticket=$(echo $resp | jq -r .ticket)

resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/protocol/openid-connect/token" \
  -H "Content-Type: application/x-www-form-urlencoded" \
  -H "Authorization: Bearer $userToken" \
  -d "audience=$KC_CLIENT" \
  -d "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" \
  -d "response_include_resource_name=true" \
  -d "ticket=$ticket")
rpt2=$(echo $resp | jq -r .access_token)

echo "Permissions using Ticket:"
echo $rpt2 | jq -R 'split(".") | .[0],.[1] | @base64d | fromjson | .authorization.permissions'
```

**Comment by pedroigor** [comment]
@sirkrypt0 Looks the difference here is that when you are querying all permissions you don't have any policy granting access to the resource, even though it is owned by the user.

However, when you are using a ticket, the server makes the decision implicitly based on the resource owner.

I see the issue you are presenting here but that is how it should work. We could change the `ResourcePolicyProvider` to also take into account the resource ownership and if the same as the subject making the request we grant access.

But yeah, the behavior is expected.

**Comment by pedroigor** [comment]
Thanks. I also missed the token is set to `user2` but owner is hardcoded to `user1`.

We are in a similar situation. Please see https://github.com/keycloak/keycloak/blob/d3b43a9f59764b763bffc11ba1f087f929f433ef/server-spi-private/src/main/java/org/keycloak/authorization/permission/Permissions.java#L63.

That method is what we use to resolve the resources whose access should be evaluated. Note that we'll only include that resource to the list if there is a ticket `granted` to the user for that resource. As you are just creating the ticket, without granting access (resource owner approving access), the resource is not added and the corresponding permissions are not returned.

When using a ticket it works because in this case you are sending the resource id and the permissions for the resource is evaluated based on the default permissions.

Makes sense now?

**Comment by sirkrypt0** [comment]
Alright, so the permission is not included in the first response, as the owner `user1` hasn't explicitly granted access to `user2`, and so the permissions for this resource are not evaluated.
But, when asking for the resource directly using a ticket in request 2, the permissions are evaluated and as the default permissions grant access, the permission is included in the response.

I had expected that simply requesting a token (without a ticket) would return all resources to which a user has access. I also looked at the docs [Authorization Services: Obtaining permissions](https://www.keycloak.org/docs/latest/authorization_services/index.html#_service_obtaining_permissions), in which it says:

>  Example of a authorization request when a client is seeking access to any resource and scope protected by a resource server.
> `curl -X POST http://${host}:${port}/realms/${realm}/protocol/openid-connect/token -H "Authorization: Bearer ${access_token}" --data "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket" --data "audience={resource_server_client_id}"`

Would evaluating the permissions for all resources be against the UMA standard or is that just for performance reasons?

**Comment by pedroigor** [comment]
Yeah ... Performance. Fetching all permissions is expensive and that is one limitation. By relying on the permission tickets we can fetch the resources faster without going over all the resources in the database.

Note that the "all permission" operation is not UMA but KC-specific.

**Comment by sirkrypt0** [comment]
Got it, thanks for your quick replies!

I submitted [this PR](https://github.com/keycloak/keycloak-documentation/pull/1580) which might make the example that confused me clearer. Feel free to change or close it if you disagree :)
